File: python/mirage/core/postgres/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import hashlib

import orjson

from mirage.accessor.postgres import PostgresAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.postgres import _client
from mirage.core.postgres.scope import detect_scope
from mirage.types import FileStat, FileType, PathSpec
from mirage.utils.errors import enoent


async def stat(accessor: PostgresAccessor,
               path: PathSpec,
               index: IndexCacheStore = None) -> FileStat:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    prefix = path.prefix
    raw = path.original
    if prefix and raw.startswith(prefix):
        raw = raw[len(prefix):] or "/"
    scope = detect_scope(PathSpec(original=raw, directory=raw, prefix=prefix))

    if scope.level == "root":
        return FileStat(name="/", type=FileType.DIRECTORY)

    if scope.level == "database_json":
        return FileStat(name="database.json", type=FileType.JSON)

    if scope.level == "schema":
        if not await _schema_exists(accessor, scope.schema):
            raise enoent(path)
        return FileStat(name=scope.schema,
                        type=FileType.DIRECTORY,
                        extra={"schema": scope.schema})

    if scope.level == "kind":
        if not await _schema_exists(accessor, scope.schema):
            raise enoent(path)
        return FileStat(name=scope.kind,
                        type=FileType.DIRECTORY,
                        extra={
                            "schema": scope.schema,
                            "kind": scope.kind
                        })

    if scope.level == "entity":
        if not await _entity_exists(accessor, scope.schema, scope.kind,
                                    scope.entity):
            raise enoent(path)
        return FileStat(name=scope.entity,
                        type=FileType.DIRECTORY,
                        extra={
                            "schema": scope.schema,
                            "kind": scope.kind,
                            "name": scope.entity
                        })

    if scope.level == "entity_schema":
        if not await _entity_exists(accessor, scope.schema, scope.kind,
                                    scope.entity):
            raise enoent(path)
        return FileStat(name="schema.json",
                        type=FileType.JSON,
                        extra={
                            "schema": scope.schema,
                            "kind": scope.kind,
                            "name": scope.entity
                        })

    if scope.level == "entity_rows":
        if not await _entity_exists(accessor, scope.schema, scope.kind,
                                    scope.entity):
            raise enoent(path)
        return await _rows_stat(accessor, scope.schema, scope.kind,
                                scope.entity)

    raise enoent(path)


async def _schema_exists(accessor: PostgresAccessor, schema: str) -> bool:
    pool = await accessor.pool()
    async with pool.acquire() as conn:
        schemas = await _client.list_schemas(conn, accessor.config.schemas)
    return schema in schemas


async def _entity_exists(accessor: PostgresAccessor, schema: str, kind: str,
                         entity: str) -> bool:
    pool = await accessor.pool()
    async with pool.acquire() as conn:
        if kind == "tables":
            names = await _client.list_tables(conn, schema)
        else:
            views = await _client.list_views(conn, schema)
            mviews = await _client.list_matviews(conn, schema)
            names = sorted(set(views) | set(mviews))
    return entity in names


async def _rows_stat(accessor: PostgresAccessor, schema: str, kind: str,
                     entity: str) -> FileStat:
    pool = await accessor.pool()
    async with pool.acquire() as conn:
        cols = await _client.fetch_columns(conn, schema, entity)
        rows = await _client.estimated_row_count(conn, schema, entity)
        size = await _client.table_size_bytes(conn, schema, entity)
    fp_payload = orjson.dumps({"columns": cols, "rows": rows})
    fingerprint = hashlib.sha256(fp_payload).hexdigest()
    return FileStat(
        name="rows.jsonl",
        type=FileType.TEXT,
        size=size,
        fingerprint=fingerprint,
        extra={
            "schema": schema,
            "kind": kind,
            "name": entity,
            "row_count": rows,
            "size_bytes": size
        },
    )

File: typescript/packages/core/src/core/postgres/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { FileStat, FileType, PathSpec } from '../../types.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { sha256Hex } from '../../utils/hash.ts'
import type { PostgresAccessor } from '../../accessor/postgres.ts'
import {
  estimatedRowCount,
  fetchColumns,
  listMatviews,
  listSchemas,
  listTables,
  listViews,
  tableSizeBytes,
} from './_client.ts'
import { detectScope } from './scope.ts'
import { enoent } from '../../utils/errors.ts'

async function schemaExists(accessor: PostgresAccessor, schema: string): Promise<boolean> {
  const schemas = await listSchemas(accessor, accessor.config.schemas)
  return schemas.includes(schema)
}

async function entityExists(
  accessor: PostgresAccessor,
  schema: string,
  kind: string,
  entity: string,
): Promise<boolean> {
  let names: string[]
  if (kind === 'tables') {
    names = await listTables(accessor, schema)
  } else {
    const views = await listViews(accessor, schema)
    const mviews = await listMatviews(accessor, schema)
    names = [...new Set([...views, ...mviews])]
  }
  return names.includes(entity)
}

export async function stat(
  accessor: PostgresAccessor,
  path: PathSpec | string,
  _index?: IndexCacheStore,
): Promise<FileStat> {
  const spec = typeof path === 'string' ? PathSpec.fromStrPath(path) : path
  const prefix = spec.prefix
  let raw = spec.original
  if (prefix !== '' && raw.startsWith(prefix)) {
    raw = raw.slice(prefix.length) || '/'
  }
  const scope = detectScope(new PathSpec({ original: raw, directory: raw, prefix }))

  if (scope.level === 'root') {
    return new FileStat({ name: '/', type: FileType.DIRECTORY })
  }
  if (scope.level === 'database_json') {
    return new FileStat({ name: 'database.json', type: FileType.JSON })
  }
  if (scope.level === 'schema') {
    if (!(await schemaExists(accessor, scope.schema))) throw enoent(spec)
    return new FileStat({
      name: scope.schema,
      type: FileType.DIRECTORY,
      extra: { schema: scope.schema },
    })
  }
  if (scope.level === 'kind') {
    if (!(await schemaExists(accessor, scope.schema))) throw enoent(spec)
    return new FileStat({
      name: scope.kind,
      type: FileType.DIRECTORY,
      extra: { schema: scope.schema, kind: scope.kind },
    })
  }
  if (scope.level === 'entity') {
    if (!(await entityExists(accessor, scope.schema, scope.kind, scope.entity))) throw enoent(spec)
    return new FileStat({
      name: scope.entity,
      type: FileType.DIRECTORY,
      extra: { schema: scope.schema, kind: scope.kind, name: scope.entity },
    })
  }
  if (scope.level === 'entity_schema') {
    if (!(await entityExists(accessor, scope.schema, scope.kind, scope.entity))) throw enoent(spec)
    return new FileStat({
      name: 'schema.json',
      type: FileType.JSON,
      extra: { schema: scope.schema, kind: scope.kind, name: scope.entity },
    })
  }
  if (scope.level === 'entity_rows') {
    if (!(await entityExists(accessor, scope.schema, scope.kind, scope.entity))) throw enoent(spec)
    return rowsStat(accessor, scope.schema, scope.kind, scope.entity)
  }
  throw enoent(spec)
}

async function rowsStat(
  accessor: PostgresAccessor,
  schema: string,
  kind: string,
  entity: string,
): Promise<FileStat> {
  const cols = await fetchColumns(accessor, schema, entity)
  const rows = await estimatedRowCount(accessor, schema, entity)
  const size = await tableSizeBytes(accessor, schema, entity)
  const fingerprint = await sha256Hex(
    new TextEncoder().encode(JSON.stringify({ columns: cols, rows })),
  )
  return new FileStat({
    name: 'rows.jsonl',
    type: FileType.TEXT,
    size,
    fingerprint,
    extra: {
      schema,
      kind,
      name: entity,
      row_count: rows,
      size_bytes: size,
    },
  })
}
